test_common/harness/kernelHelpers.cpp

```diff
@@ -809,6 +810,71 @@ int create_single_kernel_helper(cl_context context,
                                 const char *buildOptions,
                                 const bool openclCXX)
 {
+
+    // Check the build options for the -cl-std option.
+    if (!buildOptions || !strstr(buildOptions, "-cl-std"))
+    {
+        // If the build option isn't present add it using the latetest OpenCL-C
+        // version supported by the device. This allows calling code to force a
+        // particular CL C version if it is required, but also means that
+        // callers need not specify a version if they want to assume the most
+        // recent CL C.
+
+        // First get all the devices in the context and find the minumum
+        // supported OpenCL C version.
+        size_t devices_size_in_bytes{};
+        auto error = clGetContextInfo(context, CL_CONTEXT_DEVICES, 0, nullptr,
+                                      &devices_size_in_bytes);
+        test_error(error, "clGetDeviceInfo failed for CL_CONTEXT_DEVICES");
+        std::vector<cl_device_id> devices(devices_size_in_bytes
+                                          / sizeof(cl_device_id));
+        error =
+            clGetContextInfo(context, CL_CONTEXT_DEVICES, devices_size_in_bytes,
+                             devices.data(), nullptr);
+        auto current_version = get_device_cl_version(devices[0]);
+        std::for_each(std::next(devices.begin()), devices.end(),
+                      [&current_version](cl_device_id device) {
+                          auto device_version = get_device_cl_version(device);
+                          // OpenCL 3.0 is not backwards compatible with 2.0.
+                          // If we have 3.0 and 2.0 in the same driver we
+                          // use 1.2.
+                          if (((device_version >= Version(2, 0)
+                                && device_version < Version(3, 0))
+                               && current_version >= Version(3, 0))
+                              || (device_version >= Version(3, 0)
+                                  && (current_version >= Version(2, 0)
+                                      && current_version < Version(3, 0))))
+                          {
+                              current_version = Version(1, 2);
+                          }
+                          else
+                          {
+                              current_version =
+                                  (std::min)(device_version, current_version);
+                          }
+                      });
+
+        std::string cl_std{};
+        if (current_version == Version(3, 0))
```

Comment:
I'm not sure how we can future-proof this part (or if we even need to at the moment) as we don't know if/when we might get a new CL C version.

The worst that happens in this case is that we fall back to 1.2 so shouldn't introduce any problems.

Reply:
Yeah, this is a good point. It's a bit tricky since we are using the device OpenCL version to determine the OpenCL C version which isn't a 1-1 mapping and it could change with future versions like you pointed out. We could change line 858 to `current_version >= Version(3, 0)`, that way assuming future OpenCL C versions are compatible with OpenCL 3.0 we would at least get the OpenCL 3.0 language features? But I think this logic would probably need to get updated for future OpenCL C versions anyway.

Reply:
I've updated this to use `current_version >= Version(3, 0)`.